neuraxle/checkpoints.py

```diff
@@ -19,185 +19,558 @@
+
+    def save_checkpoint(self, data_container: DataContainer, context: ExecutionContext) -> DataContainer:
         """
-        Whether or not we should resume the pipeline (if the checkpoint exists)
+        Save data container data inputs with :py:attr:`~data_input_checkpointer`.
+        Save data container expected outputs with :py:attr:`~expected_output_checkpointer`.
 
-        :param context: execution context
-        :param data_container: data to resume
+        :param data_container: data container to checkpoint
+        :type data_container: DataContainer
+        :param context: execution context to checkpoint from
+        :type context: ExecutionContext
         :return:
         """
-        self.set_checkpoint_path(context.get_path())
-        return self._checkpoint_exists(data_container)
+        for current_id, data_input, _ in data_container:
+            self.data_input_checkpointer.save(current_id=current_id, data=data_input, checkpoint_path=context.get_path())
+
+        if self.expected_output_checkpointer is None:
+            if context.get_execution_mode() == ExecutionMode.FIT or context.get_execution_mode() == ExecutionMode.FIT_TRANSFORM:
+                warnings.warn('Saving Data Checkpoint {0} without an expected output checkpointer with Step Saver {1}. Perhaps, you wanted to have an expected output checkpointer ?'.format(context.get_path(), self.__class__.__name__))
+            return data_container
+
+        for current_id, _, expected_output in data_container:
+            self.expected_output_checkpointer.save(current_id=current_id, data=expected_output, checkpoint_path=context.get_path())
 
-    def _checkpoint_exists(self, data_container: DataContainer) -> bool:
+        return data_container
+
+    def read_checkpoint(self, data_container: DataContainer, context: ExecutionContext) -> DataContainer:
         """
-        Returns True if the checkpoints for each data input id exists
-        :param data_container:
-        :return:
+        Read data container data inputs checkpoint with :py:attr:`~data_input_checkpointer`.
+        Read data container expected outputs checkpoint with :py:attr:`~expected_output_checkpointer`.
+
+        :param data_container: data container to read checkpoint for
+        :type data_container: DataContainer
+        :param context: execution context to read checkpoint from
+        :type context: ExecutionContext
+        :return: data container checkpoint
+        :rtype: DataContainer
         """
+        data_inputs = []
+        for current_id in data_container.current_ids:
+            checkpoint = self.data_input_checkpointer.read(
+                current_id=current_id,
+                checkpoint_path=context.get_path()
+            )
+            data_inputs.append(checkpoint)
+        data_container.set_data_inputs(data_inputs)
+
+        if self.expected_output_checkpointer is None:
+            data_container.set_expected_outputs(None)
```

Comment:
I see, we set it to one None instead of a list of Nones as long as we have IDs. IDK if the null object can handle this. If not, then we can't do the null object as it would serve no better purpose.

Reply:
We could merge the two for loops into one perhaps, which would make the method 3x simpler.

Reply:
The save method too, not just the current load.

Reply:
done